### PR TITLE
fix(fetch-more): prevent fetch more from breaking after a non-graphql error

### PR DESCRIPTION
**Pull request by sondremare**
When using fetch more, the original query is set as the fetchMoreForQueryId.
In the old solution, when a non-graphql error occurred, both the query to fetch more and the original query were marked as erroneous. Thus both of them were removed from the store. This left the application in a weird state with null pointers and other unwanted side effects.

In this new solution we mark the fetch-more query as erroneous, but we set the original query as completed. Because the original query has already been fetched, and should not be invalidated/removed when the fetch-more query fails.

### CODE REVIEW

**Comment by sondremare** [comment]
I am not 100% sure if this is the way to go. But it is atleast an attempt. I do however need help with the tests.

**Comment by sondremare** [comment]
Fixes the problem that is referred to in https://github.com/apollographql/apollo-client/pull/2535 and https://github.com/apollographql/apollo-client/issues/2539,

**Comment by seanyu4296** [comment]
Hi im also facing this issue. I tried patching queries.js with your fix directly in my apollo-client. But im facing a weird issue, apollo client seems to call fetchMore alot of times then just breaks the whole store @sondremare . Did it work for you?

**Comment by sondremare** [comment]
It works for me yes. What kind of errors are you receiving @seanyu4296?

**Comment by seanyu4296** [comment]
I'm currently using:
react-apollo: 2.0.4
apollo-client: 2.2.0
apollo-link-http: 1.3.2
apollo-link-dedup: 1.0.5
apollo-link: 1.0.7

im trying to apply the patch through going directly to queries.js then replacing your changes
```
  QueryStore.prototype.markQueryError = function (queryId, error, fetchMoreForQueryId) {
    if (!this.store[queryId]) return;
    this.store[queryId].networkError = error;
    this.store[queryId].networkStatus = NetworkStatus.error;
    if (typeof fetchMoreForQueryId === 'string') {
      // this.markQueryError(fetchMoreForQueryId, error, undefined);
      this.markQueryResultClient(fetchMoreForQueryId, true);
    }
  };
``` 
is this correct?
@sondremare 
Did you try applying the HOC fix mentioned here by  wdimiceli regarding the refetch i think?

**Comment by sondremare** [comment]
I have not tried the HOC fix no. In general, not a fan of patching stuff. I would rather like to fix it at the source. What you have done looks correct. But you might need to change in bundle.umd.js instead. That's atleast the file that's being used in my react project. You can easily check by adding a console.log to queries.js and to bundle.umd.js to see which is being run @seanyu4296

**Comment by alewiahmed** [comment]
hey @sondremare , thank you very much. This works. One thing I can add is, I was using react apollo 2.0.4 and your fix didn't work. But after I upgraded to 2.1.0.beta.0 it's working fine. I think your fix still needs https://github.com/apollographql/react-apollo/pull/1531 to work.

**Comment by seanyu4296** [comment]
After applying the patch in bundle.umd.js and upgrading react-apollo to the current 2.1.0 beta your PR seems to have solved the issue @sondremare. I hope the maintainers can take a look at your solution and merge it soon

**Comment by jbaxleyiii** [comment]
@sondremare thanks for taking a look into this! Would you be able to take a look into the failing tests?

**Comment by sondremare** [comment]
@jbaxleyiii sorry for late reply. Been busy. To be honest, I tried fixing the tests but they were written in a way so that I was unable to fix them. Need some help here

**Comment by sondremare** [comment]
@jbaxleyiii Any chance of helping out here? Not great at writing tests for observables

**Comment by sondremare** [comment]
@jbaxleyiii I have now corrected the tests. But for some reason, subscribeToMore times on Travis, but works perfectly fine locally...

**Comment by duner** [comment]
Hi @sondremare: I had a question about this change.

Are there any examples of how to check for an error in fetchMore differently from an error to the query?It makes sense to me that an error in a fetchMore shouldn't error out the entire query. Our use case is pagination and we definitely wouldn't want the whole list of stories already loaded to show an error if the next page is unable to load. However, we _would_ want to be able to show an error at the bottom of the list of items.

tldr: In your note on this PR you said the "new solution we mark the fetch-more query as erroneous." How do we check for that error? (If you're able to give specific context to react-apollo that would be awesome)

For reference, we're using `"react-apollo": "^2.1.3"` and `"apollo-client": "^2.2.8"`.

**Comment by seanyu4296** [comment]
Hi just want to bump in @duner and try to help. I also use fetchMore for pagination and I'm able to show the fetchMore error through just saving the state in the component I'm using

Ex. 
```
fetchMore().catch((e) => this.setState({ fetchMoreErr: e}) ;

**Comment by duner** [comment]
@seanyu4296 Sorry for the delayed response — I got sucked into some other work. That makes a lot of sense. Thanks so much for the help!
